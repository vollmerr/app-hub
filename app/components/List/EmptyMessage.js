import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { DefaultButton } from 'office-ui-fabric-react/lib/Button';


export const Wrapper = styled.div`
  display: inline-flex;
  flex-direction: column;
  align-items: center;
`;


export class EmptyMessage extends React.PureComponent {
  render() {
    const { message, buttonText, buttonIcon, onClick } = this.props;

    const buttonProps = {
      onClick,
      text: buttonText,
      iconProps: {
        iconName: buttonIcon,
      },
    };

    return (
      <Wrapper>
        <p>{message}</p>
        {
          buttonText &&
            <DefaultButton {...buttonProps} />
        }
      </Wrapper>
    );
  }
}


const { string, func } = PropTypes;

EmptyMessage.propTypes = {
  message: string.isRequired,
  onClick: func,
  buttonText: string,
  buttonIcon: string,
};

export default EmptyMessage;
