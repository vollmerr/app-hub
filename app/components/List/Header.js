import styled from 'styled-components';

import theme from 'utils/theme';

const Header = styled.div`
  flex: 1 ${theme.list.headerHeight}px;
  display: flex;
  padding: 5px 0;
  height: ${theme.list.headerHeight}px;
`;

export default Header;
