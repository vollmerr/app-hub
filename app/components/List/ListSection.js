import styled from 'styled-components';

import theme from 'utils/theme';
import Section from 'components/App-Content/Section';

/**
 * Contianer section for for lists
 */
const ListSection = styled(Section) `
  height: calc(
    ${(props) => props.vh || 100}vh - \
    ${theme.hub.headerHeight} - \
    ${(props) => props.margin || 0}px
  );
  min-height: 400px;
  overflow: hidden;

  .ms-Viewport {
    overflow: auto;
    flex: 10 100%;
  }

  .ms-DetailsList {
    overflow: visible;
  }

  .ms-DetailsRow-fields,
  .ms-DetailsHeader-cell {
    cursor: pointer;
  }
`;

export default ListSection;
