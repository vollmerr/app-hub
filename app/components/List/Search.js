import styled from 'styled-components';
import { SearchBox } from 'office-ui-fabric-react/lib/SearchBox';

const Search = styled(SearchBox)`
  flex: 1;
`;

export default Search;
