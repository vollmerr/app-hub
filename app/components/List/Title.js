import styled from 'styled-components';

const Title = styled.h3`
  flex: 2;
  margin: 5px;
`;

export default Title;
