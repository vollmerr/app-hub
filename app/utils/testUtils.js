import React from 'react';
import { shallow } from 'enzyme';
import 'jest-styled-components';

/**
 * Tests a styled-component for basic tests
 * such as rending correctly and adopting attributes
 * @param {Object} Component    - React component under test
 * @param {Object} extend       - React component the component under test extends /  inherits from
 */
export function testStyledComponent(Component, extend = null) {
  describe(`<${Component.displayName} />`, () => {
    let wrapper;
    beforeEach(() => {
      wrapper = shallow(<Component />);
    });

    it('should render correctly', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it(`should render an <${Component.target}> tag`, () => {
      expect(wrapper.type()).toEqual(Component.target);
    });

    it('should have a className attribute', () => {
      expect(wrapper.prop('className')).toBeDefined();
    });

    it('should adopt a valid attribute', () => {
      const id = 'test';
      wrapper.setProps({ id });
      expect(wrapper.prop('id')).toEqual(id);
    });

    if (extend) {
      it(`should extend ${extend.name}`, () => {
        expect(wrapper.find(extend).exists()).toEqual(true);
      });
    } else {
      it('should not adopt an invalid attribute', () => {
        const attribute = 'test';
        wrapper.setProps({ attribute });
        expect(wrapper.prop('attribute')).toBeUndefined();
      });
    }
  });
}

/**
 * Tests a components for basic mapDispatchToProps information
 * @param {func} mapDispatchToProps   - components mapDispatchToProps function
 * @param {object} actions            - functions in mapDispatchToProps to test
 */
export function testMapDispatchToProps(mapDispatchToProps, actions) {
  const onAction = (action) => (
    `on${action.charAt(0).toUpperCase()}${action.slice(1)}`
  );

  let dispatch;
  let dispatchProps;
  beforeEach(() => {
    dispatch = jest.fn();
    dispatchProps = mapDispatchToProps(dispatch);
  });

  Object.keys(actions).forEach((action) => {
    describe(onAction(action), () => {
      it('should be injected', () => {
        expect(dispatchProps[onAction(action)]).toBeDefined();
      });

      it('should dispatch when called', () => {
        dispatchProps[onAction(action)]();
        expect(dispatch).toHaveBeenCalled();
      });
    });
  });
}
