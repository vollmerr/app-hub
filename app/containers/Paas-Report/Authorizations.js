import React from 'react';
import styled from 'styled-components';

import List from 'components/List';
import theme from 'utils/theme';


export const StyledList = styled(List)`
  flex: 1;
  margin: ${theme.hub.padding / 2}px;
`;


class Authorizations extends React.PureComponent {
  render() {
    return (
      <StyledList {...this.props} />
    );
  }
}


export default Authorizations;
