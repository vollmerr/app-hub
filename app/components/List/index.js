import React from 'react';
import PropTypes from 'prop-types';
import { DetailsList, SelectionMode } from 'office-ui-fabric-react/lib/DetailsList';
import { TooltipHost } from 'office-ui-fabric-react/lib/Tooltip';
import orderBy from 'lodash/orderBy';
import isEqual from 'lodash/isEqual';

import { escapeRegExp } from 'utils/string';
import { formatItem } from 'utils/data';

import Wrapper from './Wrapper';
import Title from './Title';
import Search from './Search';
import Header from './Header';
import EmptyMessage from './EmptyMessage';


/**
 * Helper function to handle selecting a single item from a list
 * Unselects the item so it can immediately be selected again
 *
 * Must be bound to a component to update `selectedItem` to the
 * item selected. Calls the callback with item if passed one.
 *
 * @param {object} component    - component to set state on (will be `this`)
 * @param {object} selection    - instance of `Selection` to use
 * @param {func} callback       - called if item is selected
 */
export const handleSelectItem = (component, selection, callback) => {
  const count = selection.getSelectedCount();
  // default to object so no error in modal
  let item = {};
  // if there are any selected items
  if (count) {
    // get the item from the selction object
    item = selection.getSelection()[0];
    // get the index of the selected item
    const index = selection.getSelectedIndices()[0];
    // unselect that item, so it can be reselected later
    selection.toggleKeySelected(index);
  }
  // set that item as selected, do callback if one is actually selected
  component.setState({ selectedItem: item }, () => (
    callback && Object.keys(item).length ? callback(item) : null
  ));
};


/**
 * List for displaying table-like data. Extends office-ui-fabric-react's
 * DetailsList, adding sorting, filtering, search box, title, an empty
 * message, and formatted item rendering.
 *
 * @return {JSX}          - List of data or empty message
 */
class List extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      items: props.items,
      columns: this.bindColumnClick(props.columns),
      sortOrder: props.columns.map((col) => col.key),
      searchValue: '',
    };
  }

  componentDidMount() {
    const { sortBy } = this.props;
    if (sortBy) {
      sortBy.forEach((x) => {
        const column = this.state.columns.find((col) => col.key === x);
        this.handleColumnClick(null, column);
      });
    }
  }

  componentWillReceiveProps(nextProps) {
    const { items } = this.props;
    // different items, updae local ones (deep compare...)
    if (!isEqual(items, nextProps.items)) {
      this.setState({
        items: nextProps.items,
        columns: this.bindColumnClick(nextProps.columns),
        sortOrder: nextProps.columns.map((col) => col.key),
        searchValue: '',
      });
    }
  }

  /**
   * Binds the method for clicking on a column header to each column
   */
  bindColumnClick = (columns) => (
    columns.map((col) => ({
      ...col,
      ...!col.notSortable && { onColumnClick: this.handleColumnClick },
    }))
  );

  /**
   * Handles filtering list items displayed based off value
   *
   * @param {string} value         - value to filter results by
   */
  handleSearch = (value) => {
    const re = new RegExp(escapeRegExp(value), 'i');
    const items = this.props.items.filter((item) => (
      Object.values(item).some((x) => re.test(x))
    ));
    this.setState({ items, searchValue: value });
  }

  /**
   * Handles clicking on a column header
   *
   * @param {event} event         - click event on column header
   * @param {object} column       - column clicked on
   */
  handleColumnClick = (event, column) => {
    const { columns } = this.state;
    const selected = columns.findIndex((col) => col.key === column.key);
    const newColumns = [...columns];

    // 3 way state -> desc (true), asc (false), not sorted (undefined)
    const isDesc = columns[selected].isSortedDescending;
    if (isDesc === undefined || isDesc) {
      newColumns[selected].isSorted = true;
      newColumns[selected].isSortedDescending = !columns[selected].isSortedDescending;
    } else {
      newColumns[selected].isSorted = false;
      newColumns[selected].isSortedDescending = undefined;
    }

    this.setState({ columns: newColumns });
    this.handleSort(newColumns, newColumns[selected].key);
  }

  /**
   * Handles sorting columns by the given sort order
   *
   * @param {object} columns      - columns of list
   * @param {string} selectedKey  - key of selected column
   */
  handleSort = (columns, selectedKey) => {
    const { items, sortOrder } = this.state;
    // move unsorted to end of list
    const sorted = [];
    const unSorted = [];
    sortOrder.forEach((key) => {
      if (columns[columns.findIndex((x) => x.key === key)].isSortedDescending === undefined) {
        unSorted.push(key);
      } else {
        sorted.push(key);
      }
    });
    // move selected to front of list
    const newSortOrder = [...sorted, ...unSorted];
    const selected = newSortOrder.findIndex((key) => key === selectedKey);
    newSortOrder.splice(selected, 1);
    newSortOrder.unshift(selectedKey);
    // build keys and isDesc for sort order
    const isDesc = newSortOrder.reduce((acc, key) => {
      const column = columns.find((col) => col.key === key);
      const desc = (column && column.isSortedDescending) ? 'desc' : 'asc';
      return [...acc, desc];
    }, []);
    // sort and update state
    const newItems = orderBy(items, newSortOrder, isDesc);
    this.setState({ items: newItems, sortOrder: newSortOrder });
  }

  /**
   * Renders the header row, displaying a tooltip if provided
   * an arialLabel
   *
   * @param {object} props        - header props
   * @param {func} defaultRender  - default render provided by DetailsList
   *
   * @return {JSX}                - column header
   */
  renderHeader = (props, defaultRender) => (
    defaultRender({
      ...props,
      onRenderColumnHeaderTooltip: (tooltipHostProps) => (
        tooltipHostProps.content ?
          <TooltipHost {...tooltipHostProps} /> :
          <div>{tooltipHostProps.children}</div>
      ),
    })
  )

  /**
   * Renders columns in a custom format
   *
   * @param {object} item     - current item (row) of list
   * @param {number} index    - index of current item
   * @param {object} column   - column to render
   *
   * @return {string}         - content to render
   */
  renderItemColumn = (item, index, column) => {
    const { enums } = this.props;
    return formatItem(item, column.fieldName, column, enums);
  }

  render() {
    const { title, empty, ...list } = this.props;
    const { items, columns, searchValue } = this.state;

    const listProps = {
      ...list,
      items,
      columns,
      onRenderDetailsHeader: this.renderHeader,
      onRenderItemColumn: this.renderItemColumn,
      onShouldVirtualize: () => false,
    };

    const searchProps = {
      value: searchValue,
      onChange: this.handleSearch,
    };

    return (
      <Wrapper>
        <Header>
          <Title>{title}</Title>
          <Search {...searchProps} />
        </Header>

        {
          items.length ?
            <DetailsList {...listProps} /> :
            <EmptyMessage {...empty} />
        }
      </Wrapper>
    );
  }
}


const { string, number, arrayOf, shape, any } = PropTypes;

List.propTypes = {
  title: string,
  items: arrayOf(any),
  columns: arrayOf(
    shape({
      key: string,
      name: string,
      fieldName: string,
    }),
  ),
  checkboxVisibility: number,
  selectionMode: number,
  empty: any,
  enums: any,
  sortBy: arrayOf(string),
};

List.defaultProps = {
  items: [],
  empty: {
    message: 'No items',
  },
  enums: {},
  selectionMode: SelectionMode.none,
};

export default List;
