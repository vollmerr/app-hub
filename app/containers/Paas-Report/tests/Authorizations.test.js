import React from 'react';
import { shallow } from 'enzyme';
import { testStyledComponent } from 'utils/testUtils';

import List from 'components/List';

import Authorizations, { StyledList } from '../Authorizations';


testStyledComponent(StyledList, List);


const props = {
  columns: [],
  items: [],
};


describe('<Authorizations />', () => {
  let wrapper;
  beforeEach(() => {
    jest.resetAllMocks();
    wrapper = shallow(<Authorizations {...props} />);
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
