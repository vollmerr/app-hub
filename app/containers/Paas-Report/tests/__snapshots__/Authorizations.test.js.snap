// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Authorizations /> should render correctly 1`] = `
<Authorizations__StyledList
  columns={Array []}
  items={Array []}
/>
`;

exports[`<Authorizations__StyledList /> should render correctly 1`] = `
.c0 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 5px;
}

<List
  className="c0"
  empty={
    Object {
      "message": "No items",
    }
  }
  enums={Object {}}
  items={Array []}
  selectionMode={0}
  style={Object {}}
/>
`;
